src/main/java/ui/UI.java

```diff
@@ -331,6 +332,84 @@ private void setupGlobalKeyboardShortcuts(Scene scene) {
         });
     }
 
+    /**
+     * Calculates initial stage dimensions using graphics environment calculations, and then
+     * attempts to set stage dimensions using screen information.
+     *
+     * @param stage The stage whose bounds will be set.
+     */
+    private void setupStageDimensions(Stage stage) {
+        // Reasonable defaults.
+        Rectangle dimensions = getDimensions();
+        stage.setMinWidth(panels.getPanelWidth());
+        stage.setMaxWidth(dimensions.getWidth());
+        stage.setMinHeight(dimensions.getHeight());
+        stage.setMaxHeight(dimensions.getHeight());
+        stage.setX(0);
+        stage.setY(0);
+        // Then calculate sizes using screen information.
+        setBoundsFromScreen(stage);
+    }
+
+    /**
+     * Configure setBoundsFromScreen to trigger whenever the x-coordinate of the given stage changes.
+     *
+     * @param stage The stage which contains the x-coordinate to listen.
+     */
+    private void setupPositionListener(Stage stage) {
+        stage.xProperty().addListener((observableValue, oldX, newX) -> setBoundsFromScreen(stage));
+    }
+
+    /**
+     * Searches for all screens that the given stage is in, and then sets the following bounds for
+     * the stage's dimensions:
+     * - Minimum height: Minimum height of the screens
+     * - Maximum height: Maximum height of the screens
+     * - Maximum width: Total width of the screens
+     *
+     * @param stage The stage whose bounds are to be set.
+     */
+    private void setBoundsFromScreen(Stage stage) {
+        List<Rectangle2D> intersectingScreenBounds = Screen.getScreensForRectangle(
+                stage.getX(),
+                stage.getY(),
+                stage.getWidth(),
+                stage.getHeight()
+        ).stream().map(Screen::getVisualBounds).collect(Collectors.toList());
+
+        if (!intersectingScreenBounds.isEmpty()) {
+            Region stageBounds = getStageBounds(intersectingScreenBounds);
+            stage.setMinHeight(stageBounds.getMinHeight());
+            stage.setMaxHeight(stageBounds.getMaxHeight());
+            stage.setMaxWidth(stageBounds.getMaxWidth());
+        }
+    }
+
+    /**
+     * Calculates a stage's boundaries based on the screen information given as argument. If the list of screens
+     * is empty, then the boundaries will have the default value of Region.USE_COMPUTED_SIZE (-1.0).
+     *
+     * @param intersectingScreenBounds A list of screens, which should be non-empty
+     * @return A Region object containing relevant information in the maxHeight, minHeight and maxWidth fields.
+     */
+    public static Region getStageBounds(List<Rectangle2D> intersectingScreenBounds) {
+        List<Double> occupyingHeights = intersectingScreenBounds.stream()
```

Comment:
Actually, using DoubleSummaryStatistics would be OK since we are already using mapToDouble. I'll change the code to use DoubleSummaryStatistics if you're fine with that :+1: